addons/emoji-picker/userscript.js

```diff
@@ -0,0 +1,551 @@
+    emojiPicker.style.top = emojiPickerOffset + "px";
+    //scratchr2 makes the body and root <html>'s height value the size of the screen somehow so I have to do this
+    let realDocumentBody =
+      addon.tab.clientVersion === "scratchr2" ? document.querySelector("#pagewrapper") : document.body;
+    if (emojiPicker.getBoundingClientRect().bottom > realDocumentBody.getBoundingClientRect().bottom - 48) {
+      //Emoji picker may be partially hidden, move up
+      emojiPicker.style.top = -emojiPicker.getBoundingClientRect().height + "px";
+    }
+  };
+
+  const showEmojiPicker = function (event) {
+    if (!event.target.classList.contains("sa-emoji-button")) return; //Only attempt to show when clicking button, not picker
+    unicodeContainer.style.display = "none";
+    pickerDivider.style.display = "none";
+
+    setSeeMoreText();
+
+    this.appendChild(emojiPicker);
+
+    setEmojiPickerPos();
+  };
+
+  const insertTextToTextArea = function (insertText, textBox) {
+    if (textBox.selectionStart === textBox.value.length && textBox.selectionEnd === textBox.value.length) {
+      //Cursor is at the end
+      if (
+        !(textBox.value[textBox.value.length - 1] === " ") &&
+        !(textBox.value[textBox.value.length - 1] === undefined)
+      ) {
+        insertText = " " + insertText;
+      }
+      textBox.value += insertText;
+    } else {
+      //Cursor is somewhere else or is selecting a part of the textbox
+      if (!(textBox.value[textBox.selectionStart] === " ")) {
+        insertText = " " + insertText;
+      }
+      if (!(textBox.value[textBox.selectionEnd - 1] === " ")) {
+        insertText += " ";
+      }
+      textBox.value =
+        textBox.value.substring(0, textBox.selectionStart) +
+        insertText +
+        textBox.value.substring(textBox.selectionEnd - 1, textBox.value.length);
+    }
+  };
+
+  document.addEventListener("mouseup", function (event) {
+    if (!emojiPicker.contains(event.target)) {
+      emojiPicker.remove();
+    }
+  });
+
+  const addEmoji = function () {
+    try {
+      var textBox;
+      if (addon.tab.clientVersion === "scratch-www") {
+        textBox =
+          this.parentElement.parentElement.parentElement.parentElement.parentElement.parentElement.querySelector(
+            'textarea[id*="frc-compose-comment"]'
+          );
+      } else {
+        //scratchr2
+        textBox = this.parentElement.parentElement.parentElement.parentElement.parentElement.querySelector(
+          ".control-group.tooltip.right > textarea"
+        );
+      }
+      insertTextToTextArea(this.dataset.text, textBox);
+      //Simulate inputting so that Scratch doesn't consider
+      //emojis added from an empty comment using this button
+      //as an empty comment
+      textBox.dispatchEvent(new Event("input", { bubbles: true }));
+    } catch (error) {}
+  };
+
+  //Addon
+
+  //Create the emoji picker
+  var emojiPicker = document.createElement("div");
```

Comment:
Why can't you use const?

Reply:
I thought you couldn't modify properties of const variables.

Reply:
You can.